scripts/migrations/4

```diff
@@ -0,0 +1,23 @@
+#!/bin/sh
+# Migration 4 --> 5
+
+helm template packages/system/fluxcd/ . -n cozy-fluxcd | kubectl delete -f - -n cozy-fluxcd || true
+
+if kubectl get crd fluxinstances.fluxcd.controlplane.io; then
+  kubectl patch crd fluxinstances.fluxcd.controlplane.io -p '{"metadata":{"finalizers":null}}' --type=merge
+fi
+if kubectl get fluxinstances.fluxcd.controlplane.io -n cozy-fluxcd flux; then
+  kubectl delete fluxinstances.fluxcd.controlplane.io -n cozy-fluxcd flux --wait=false
+  kubectl patch fluxinstances.fluxcd.controlplane.io -n cozy-fluxcd flux -p '{"metadata":{"finalizers":null}}' --type=merge
+fi
```

Comment:
I'm not sure I understand the purpose of this migration, is it just designed to evict a flux-operator that has been installed as a core package? There hasn't been a release of cozystack which had shipped flux-operator, and it doesn't look like flux is moving namespaces, I don't think that is needed.

On the other hand, it could be beneficial to move Flux to `flux-system` for compatibility with some UIs that expect Flux to be in flux-system. That would require a migration. It is not required by flux-operator though, Flux can be in any single namespace, it should be the same namespace as the operator was `helm install`'ed into.

Reply:
Installing flux-operator is designed to take over an existing Flux instance however it was deployed, so long as it's one singular instance and the namespace matches that of the `FluxInstance`.

Reply:
If the goal of this migration is to make Flux Operator user-installable inside of a tenant, I think that further discussion is needed before that will work. Note this is mentioned at the top of the guide docs:

> The Flux Operator comes with a Kubernetes CRD called [FluxInstance](https://fluxcd.control-plane.io/operator/fluxinstance/). A single custom resource of this kind can exist in a Kubernetes cluster with the name flux that must be created in the same namespace where the operator is deployed.

https://fluxcd.control-plane.io/operator/flux-config/

Reply:
I'm testing it now, if it allows the operator of cozystack to perform self-upgrades then I think it's a go. I don't know about this migration though. What will stop it from deleting FluxInstances / where is the FluxInstance to be deleted expected to come from?

Reply:
The PR as it is crafted right now appears to spin until failure:

```
yebyen@kingdon-mbp cozystack % kcs logs -f cozystack-6c958b6ffd-bw4gv -c cozystack
time="2024-06-25T22:51:23Z" level=info msg="running k8s-await-election" version=v0.4.1
time="2024-06-25T22:51:23Z" level=info msg="no status endpoint specified, will not be created"
I0625 22:51:23.847222       1 leaderelection.go:250] attempting to acquire leader lease cozy-system/cozystack...
I0625 22:51:23.856672       1 leaderelection.go:260] successfully acquired lease cozy-system/cozystack
time="2024-06-25T22:51:23Z" level=info msg="long live our new leader: 'cozystack-6c958b6ffd-bw4gv'!"
time="2024-06-25T22:51:23Z" level=info msg="starting command '/cozystack/scripts/installer.sh' with arguments: '[]'"
+ kubectl get configmap -n cozy-system cozystack -o 'go-template={{index .data "bundle-name"}}'
Error from server (NotFound): configmaps "cozystack-version" not found
configmap/cozystack-version created
make: Entering directory '/cozystack/packages/core/platform'
helm template -n cozy-system platform . --dry-run=server -a admissionregistration.k8s.io/v1 -a apiextensions.k8s.io/v1 -a apiregistration.k8s.io/v1 -a apps/v1 -a authentication.k8s.io/v1 -a authorization.k8s.io/v1 -a autoscaling/v1 -a autoscaling/v2 -a batch/v1 -a certificates.k8s.io/v1 -a coordination.k8s.io/v1 -a discovery.k8s.io/v1 -a events.k8s.io/v1 -a flowcontrol.apiserver.k8s.io/v1 -a flowcontrol.apiserver.k8s.io/v1beta3 -a networking.k8s.io/v1 -a node.k8s.io/v1 -a policy/v1 -a rbac.authorization.k8s.io/v1 -a scheduling.k8s.io/v1 -a storage.k8s.io/v1 -a v1 -s templates/namespaces.yaml | kubectl apply -n cozy-system -f-
namespace/cozy-cert-manager created
namespace/cozy-cilium created
namespace/cozy-clickhouse-operator created
namespace/cozy-cluster-api created
namespace/cozy-dashboard created
namespace/cozy-etcd-operator created
namespace/cozy-fluxcd created
namespace/cozy-grafana-operator created
namespace/cozy-kafka-operator created
namespace/cozy-kamaji created
namespace/cozy-kubeovn created
namespace/cozy-kubevirt created
namespace/cozy-kubevirt-cdi created
namespace/cozy-linstor created
namespace/cozy-mariadb-operator created
namespace/cozy-metallb created
namespace/cozy-monitoring created
namespace/cozy-postgres-operator created
namespace/cozy-public created
namespace/cozy-rabbitmq-operator created
namespace/cozy-redis-operator created
namespace/cozy-telepresence created
namespace/cozy-victoria-metrics-operator created
make: Leaving directory '/cozystack/packages/core/platform'
Error from server (NotFound): deployments.apps "fluxcd-flux-operator" not found
make: Entering directory '/cozystack/packages/system/fluxcd-operator'
helm upgrade -i -n cozy-fluxcd fluxcd-operator . -f -
Release "fluxcd-operator" does not exist. Installing it now.
W0625 22:51:24.966602     171 warnings.go:70] would violate PodSecurity "restricted:latest": host namespaces (hostNetwork=true)
NAME: fluxcd-operator
LAST DEPLOYED: Tue Jun 25 22:51:24 2024
NAMESPACE: cozy-fluxcd
STATUS: deployed
REVISION: 1
TEST SUITE: None
make: Leaving directory '/cozystack/packages/system/fluxcd-operator'
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmrepositories.source.toolkit.fluxcd.io" not found
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "helmcharts.source.toolkit.fluxcd.io" not found
Terminated
Flux CRDs are not ready
time="2024-06-25T22:52:24Z" level=fatal msg="failed to run" err="exit status 1"
```

I am testing removing the migration to see if it does fix the issue - I was afraid that migration would do this.

Reply:
Hey @kingdonb I reworkied this fully, could you check again please?

Reply:
Oh, you're right I did something wrong here.

Hey, we didn't make any release since flux-operator was merged, let's remove it.